### PR TITLE
APIS-5849 - Refactor role/request in preparation for new LDAP user access

### PR DESCRIPTION
**Pull request by AndySpaven**

### CODE REVIEW

File: app/controllers/TeamMembersController.scala [hunk 1]
```diff
@@ -41,11 +41,11 @@ import services.{ApplicationService, ApmService, DeveloperService}
 trait WithRestrictedApp {
   self: TeamMembersController =>
 
-  def withRestrictedApp(appId: ApplicationId)(f: ApplicationWithHistory => Future[Result])(implicit request: LoggedInRequest[_], ec: ExecutionContext, hc: HeaderCarrier, strideAuthConfig: StrideAuthConfig) = {
+  def withRestrictedApp(appId: ApplicationId)(f: ApplicationWithHistory => Future[Result])(implicit request: LoggedInRequest[_], ec: ExecutionContext, hc: HeaderCarrier) = {
     withApp(appId) { app =>
       app.application.access match {
         case _: Standard => f(app)
-        case _ if isAtLeastSuperUser => f(app)
+        case _ if request.role.isSuperUser => f(app)
```
Comment (peteslater-ee) [inline app/controllers/TeamMembersController.scala:48]: I think we may have lost some precision here :(

File: app/modules/stride/utils/GatekeeperAuthorisationHelper.scala [hunk 2]
```diff
@@ -17,17 +17,17 @@
...
 import uk.gov.hmrc.modules.stride.domain.models.LoggedInUser
+// import uk.gov.hmrc.modules.stride.domain.models.GatekeeperRole
 
 trait GatekeeperAuthorisationHelper {
   implicit def loggedIn(implicit request: LoggedInRequest[_]): LoggedInUser = LoggedInUser(request.name)
   
-  def isAtLeastSuperUser(implicit request: LoggedInRequest[_], strideAuthConfig: StrideAuthConfig): Boolean = {
-    request.authorisedEnrolments.getEnrolment(strideAuthConfig.superUserRole).isDefined || request.authorisedEnrolments.getEnrolment(strideAuthConfig.adminRole).isDefined
-  }
+  // def request.role.isSuperUser(implicit request: LoggedInRequest[_]): Boolean = {
```
Comment (peteslater-ee) [inline app/modules/stride/utils/GatekeeperAuthorisationHelper.scala:26]: Remove commented out stuffs

File: app/controllers/TeamMembersController.scala (hunk 1 above)
Comment (AndySpaven) [reply app/controllers/TeamMembersController.scala:48]: No - it's exactly the same - just now an ADMIN counts as a SuperUser too.

File: app/modules/stride/controllers/actions/GatekeeperAuthorisationActions.scala [hunk 3]
```diff
@@ -66,7 +66,16 @@ trait GatekeeperAuthorisationActions {
         val retrieval = Retrievals.name and Retrievals.authorisedEnrolments
 
         authConnector.authorise(predicate, retrieval) map {
-          case Some(name) ~ authorisedEnrolments => Right(new LoggedInRequest(name.name, authorisedEnrolments, request))
+          case Some(name) ~ authorisedEnrolments => 
+            def applyRole(role: GatekeeperRole): Either[Result, LoggedInRequest[A]] =  Right(new LoggedInRequest(name.name, role, request))
+
+            ( authorisedEnrolments.getEnrolment(strideAuthConfig.adminRole).isDefined, authorisedEnrolments.getEnrolment(strideAuthConfig.superUserRole).isDefined, authorisedEnrolments.getEnrolment(strideAuthConfig.userRole).isDefined ) match {
```
Comment (robdawson-hmrc) [inline app/modules/stride/controllers/actions/GatekeeperAuthorisationActions.scala:72]: is there any significant cost (eg db/network call ) to calling authorisedEnrolments.getEnrolment? we always call it 3 times here even if the 'adminRole' value returns true making the other calls redundant

File: app/controllers/EmailsController.scala [hunk 4]
```diff
@@ -25,7 +25,7 @@ import model.DeveloperStatusFilter.VerifiedStatus
 import model.EmailOptionChoice.{EMAIL_ALL_USERS, _}
 import model.EmailPreferencesChoice.{SPECIFIC_API, TAX_REGIME, TOPIC}
 import model.TopicOptionChoice.TopicOptionChoice
-import model.{APIAccessType, APICategory, AnyEnvironment, ApiContextVersion, ApiDefinition, CombinedApi, Developers2Filter, DropDownValue, EmailOptionChoice, GatekeeperRole, RegisteredUser, SendEmailChoice, SendEmailPreferencesChoice, TopicOptionChoice, User}
+import model._
 import model._
```
Comment (robdawson-hmrc) [inline app/controllers/EmailsController.scala:29]: duplicated imports

File: app/modules/stride/controllers/actions/GatekeeperAuthorisationActions.scala (hunk 3 above)
Comment (peteslater-ee) [reply app/modules/stride/controllers/actions/GatekeeperAuthorisationActions.scala:72]: Nope - enrolments is created with a set of enrolments and these are just looking for an entry in that set.
So no db or network overhead for each lookup.

**Pull request merged**